### PR TITLE
[5.1] Intuitive use for more common Arr:: and Collection::first() pattern

### PR DESCRIPTION
**Pull request by evictor**

### CODE REVIEW

**Comment by taylorotwell** [comment]
Need description on what you're actually trying to add here.

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ composer.lock
 .php_cs.cache
 .DS_Store
 Thumbs.db
+.idea/
```
Comment (GrahamCampbell) [inline .gitignore]: please don't add this

File: src/Illuminate/Support/Arr.php [hunk 2]
```diff
@@ -149,8 +149,12 @@ public static function fetch($array, $key)
      */
     public static function first($array, callable $callback, $default = null)
     {
+        $reflection = new \ReflectionFunction($callback);
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: import

**Comment by evictor** [comment]
Fix style analysis + import for @GrahamCampbell

File: .gitignore (hunk 1 above)
Comment (evictor) [inline .gitignore]: Why not? It hurts nothing... Only is useful for people who use a certain IDE, just like how there are ignores intended for people who use a certain OS, etc.

Comment (joecohens) [inline .gitignore]: Imagine the mess if we add every IDE hidden files to the gitignore, maybe just add it to your global gitignore?

**Comment by evictor** [comment]
@taylorotwell currently in order to use `::first()` with a callback, your callback must take two params `$key` and `$value` regardless of whether it uses `$key`. This allows usage of a 1-param callback for the purpose of evaluating only against values, which is perhaps a more common use case.

This is breaking for people who use `::first()` with 1-param callback to select against keys, so maybe this new approach should be renamed, e.g. to `::firstValue()`. I guess it is a valid use case, but the reason I chose reflection and put into `::first()` is because that reflects what is maybe more intuitive based on extant functional languages and the idea of array-like collections.

File: .gitignore (hunk 1 above)
Comment (evictor) [inline .gitignore]: Mess? How often do you need to peruse the ignore file? And there aren't that many IDEs... I wouldn't argue that you go through and exhaustively add all IDE ignores; but I mean if a developer comes in using a particular IDE, why not? As long as the ignore spec won't interfere with something meaningful, it will certainly never have any performance or code readability implications.

Comment (arrilot) [inline .gitignore]: A single PR should never add several unrelated changes.

Comment (evictor) [inline .gitignore]: That's fine—just omit it then. I have no emotional attachment to this particular change!

**Comment by lucasmichot** [comment]
> This is breaking for people who use ::first() with 1-param callback to select against keys

Then the PR should target `5.2`

**Comment by GrahamCampbell** [comment]
Please send to 5.2 as one commit.
